File: C#/NonBlockingExchangerSynchronizer/NonBlockingExchangerSynchronizer/TestNonBlockingExchanger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace NonBlockingExchangerSynchronizer
{
    class TestNonBlockingExchanger
    {
        static void Main(string[] args)
        {

            TestNonBlockingExchanger tests = new TestNonBlockingExchanger();

            tests.Test_Simple_Exchange_Message();
            tests.Test_Multiple_Exchange_Message();
            tests.Test_Zero_Timeout_Exchange_Message();
            tests.Test_Timeout_Exchange_Message();
            tests.Test_Interrupted_Exchange_Message();
            tests.Test_Interrupted_Rearm_Exchange_Message();

            Console.ReadKey();
        }

        public void Test_Simple_Exchange_Message()
        {
            Console.WriteLine("-------Test_Simple_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 2000)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 2000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);

            Console.WriteLine("----------------------------------------------------");

        }

        public void Test_Multiple_Exchange_Message()
        {
            Console.WriteLine("-------Test_Multiple_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 2000)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 2000)),
                   t3 = new Thread(() => ThreadFunc(exchanger,3, 2000)),
                   t4 = new Thread(() => ThreadFunc(exchanger,4, 2000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
            t4.Start();
            Thread.Sleep(500);
            
            Console.WriteLine("----------------------------------------------------");
        }

        public void Test_Zero_Timeout_Exchange_Message()
        {
            Console.WriteLine("-------Test_Zero_Timeout_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 0)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 2000)),
                   t3 = new Thread(() => ThreadFunc(exchanger,3, 2000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(1000);
            t3.Start();
            Thread.Sleep(500);
            Console.WriteLine("----------------------------------------------------");
        }

        public void Test_Timeout_Exchange_Message()
        {
            Console.WriteLine("-------Test_Timeout_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 1000)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 2000)),
                   t3 = new Thread(() => ThreadFunc(exchanger,3, 2000));

            t1.Start();
            Thread.Sleep(1500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
            Console.WriteLine("----------------------------------------------------");
        }

        public void Test_Interrupted_Exchange_Message()
        {
            Console.WriteLine("-------Test_Interrupted_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 1000)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 3000)),
                   t3 = new Thread(() => ThreadFunc(exchanger,3, 3000));

            t1.Start();
            Thread.Sleep(1000);
            t1.Interrupt();
            Thread.Sleep(20);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);

            Console.WriteLine("----------------------------------------------------");
        }

        public void Test_Interrupted_Rearm_Exchange_Message()
        {
            Console.WriteLine("-------Test_Interrupted_Rearm_Exchange_Message - BEGIN--------");
            NonBlockingExchanger<string> exchanger = new NonBlockingExchanger<string>();
            Thread t1 = new Thread(() => ThreadFunc(exchanger,1, 1000)),
                   t2 = new Thread(() => ThreadFunc(exchanger,2, 3000)),
                   t3 = new Thread(() => ThreadFunc(exchanger,3, 3000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(10);
            t1.Interrupt();
            Thread.Sleep(20);
            t3.Start();
            Thread.Sleep(500);
            Console.WriteLine("----------------------------------------------------");
        }



        public void ThreadFunc(NonBlockingExchanger<string> nonBlockingExchanger , int id, int timeout)
        {

            String msg = "Message " + id;
            Console.WriteLine("Thread " + id + " send " + msg);
            try
            {
                msg = nonBlockingExchanger.Exchange(msg, timeout);
            }
            catch (ThreadInterruptedException)
            {
                Console.WriteLine("Thread Interrupted Exception occurred!");
                return;
            }
            catch (TimeoutException e)
            {
                Console.WriteLine(e.Message);
                return;
            }
                Console.WriteLine("Thread " + id + " received " + msg);

        }
    }
}

File: Java/NonBlockingExchangerSynchronizer/src/test/java/TestNonBlockingExchanger.java
import org.junit.Test;

/**
 * Created by Ruben Gomes on 21/06/2015.
 */
public class TestNonBlockingExchanger {

    @Test
    public void test_simple_exchange_Message() throws InterruptedException {
        System.out.println("-------test_simple_exchange_Message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,2000),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,2000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        System.out.println("-----------------------------------------");
    }

    @Test
    public void test_multiple_exchange_message() throws InterruptedException {
        System.out.println("-------test_multiple_exchange_Message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,4000),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,4000),
                t3 = new NonBlockingExchangerThread<String>(exchanger,3,4000),
                t4 = new NonBlockingExchangerThread<String>(exchanger,4,4000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        t4.start();
        Thread.sleep(500);
        System.out.println("-----------------------------------------");
    }

    @Test
    public void test_zero_timeout_exchange_message() throws InterruptedException {
        System.out.println("-------test_zero_timeout_exchange_message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,0),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,2000),
                t3 = new NonBlockingExchangerThread<String>(exchanger,3,2000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        System.out.println("-----------------------------------------");
    }

    @Test
    public void test_timeout_exchange_message() throws InterruptedException {
        System.out.println("-------test_timeout_exchange_message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,1000),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,2000),
                t3 = new NonBlockingExchangerThread<String>(exchanger,3,2000);

        t1.start();
        Thread.sleep(1500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        System.out.println("-----------------------------------------");
    }

    @Test
    public void test_interrupted_exchange_message() throws InterruptedException {
        System.out.println("-------test_interrupted_exchange_message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,1000),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,2000),
                t3 = new NonBlockingExchangerThread<String>(exchanger,3,2000);

        t1.start();
        Thread.sleep(500);
        t1.interrupt();
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        System.out.println("-----------------------------------------");
    }

    @Test
    public void test_interrupted_rearm_exchange_message() throws InterruptedException {
        System.out.println("-------test_interrupted_rearm_exchange_message - BEGIN---------");
        NonBlockingExchanger<String> exchanger = new NonBlockingExchanger<String>();
        NonBlockingExchangerThread<String>
                t1 = new NonBlockingExchangerThread<String>(exchanger,1,4000),
                t2 = new NonBlockingExchangerThread<String>(exchanger,2,4000),
                t3 = new NonBlockingExchangerThread<String>(exchanger,3,2000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(5);
        t1.interrupt();
        Thread.sleep(10);
        t3.start();
        Thread.sleep(2500);
        System.out.println("-----------------------------------------");
    }



}
